### PR TITLE
Closes #53 - invalidate all unreachable steps on forking

### LINKED ISSUE #53: Visited steps aren't forgotten if all fields in step are invalidated
If all steps in a field are invalidated the step should be removed from the steps array in sessionModel.  This would make sure that the only steps included in history are valid completed steps. It would also simplify calculating the current step number/total steps when moving through a form containing forks.

### PR DESCRIPTION
**Pull request by joefitter**
* Check for req.forkTaken on controller complete, if present then exec invalidation logic
* Get all reachable steps from path to invalidate, and path to validate, invalidate all in blacklist that aren't present in whitelist
* invalidation removes step and all step fields from sessionModel

### CODE REVIEW

**Comment by joefitter** [comment]
Linked PR - https://github.com/UKHomeOffice/passports-form-controller/pull/41

**Comment by joefitter** [comment]
@gavboulton @JoeChapman @easternbloc @daniel-ac-martin this change fixes the current issue with forking by "invalidating" unreachable steps and fields when a fork is taken - if you go back and edit a field/step causing a fork condition to be evaluated - the losing path is wiped - this takes into account any potential shared steps and leaves these - logic is:

Difference of all reachable steps from losing fork and all reachable steps from winning fork.

**Comment by gavboulton** [comment]
I haven't looked at these PRs yet but do we also need to cover the concept of branching? If forking is a two or more pathways, what happens when a user can go from a to c and some users may need to go via b1 and some via b2?

**Comment by JoeChapman** [comment]
We should probably move the discussion to an issues so we can think through the logic and expectations clearly.

**Comment by joefitter** [comment]
hey @gavboulton we cover all cases with forking logic - a fork can either add x steps to a journey, skip x steps or go down a completely different route - it all depends what you put as the target property

**Comment by joefitter** [comment]
@gavboulton @JoeChapman @daniel-ac-martin @easternbloc anyone had a chance to review this yet? Linked PR https://github.com/UKHomeOffice/passports-form-controller/pull/41

**Comment by JoeChapman** [comment]
This LGTM, @gavboulton do you see any problems with this?

**Comment by gavboulton** [comment]
Looking.

**Comment by gavboulton** [comment]
Can we have some tests please.

**Comment by joefitter** [comment]
Ah - did write a load originally in hof-controllers where this logic lived - I must have forgotten to copy them over - I'll move them now

**Comment by joefitter** [comment]
@gavboulton @JoeChapman @easternbloc @daniel-ac-martin I have added tests, also changed the logic slightly so it doesn't require a change to form-controller

**Comment by JoeChapman** [comment]
It's hard to read, but I get why. So it LGTM

**Comment by joefitter** [comment]
I know - its quite complicated logic, basically get all reachable steps from the next step (after forking conditional executed), get all reachable steps from any route that isn't taken ((fork targets + next) - next step from forking conditional), and invalidate all steps that are reachable by routes not taken, but not reachable by route taken

**Comment by joefitter** [comment]
@gavboulton is this ok to merge?

**Comment by JoeChapman** [comment]
@joefitter @gavboulton I think the tests look good, @gavboulton are you ok to merge this?

**Comment by gavboulton** [comment]
The tests describe what's actually being checked and not documenting the functionality. Could they be revised a bit please?

**Comment by joefitter** [comment]
@gavboulton not sure I follow you - could you expand a bit please?

**Comment by joefitter** [comment]
@gavboulton are you able to give an example of how you would like the tests to look?

**Comment by gavboulton** [comment]
I'll take a look.

**Comment by joefitter** [comment]
@gavboulton did you get a chance to look at this? Is it just the wording of the tests you would like changing or what is actually being tested?

**Comment by gavboulton** [comment]
@joefitter It's the wording.

**Comment by joefitter** [comment]
@gavboulton are you happy for this PR to go in?

**Comment by gavboulton** [comment]
Not yet. The wording of the tests still don't describe features.

File: test/middleware/spec.check-progress.js [hunk 1]
```diff
@@ -50,4 +51,173 @@ describe('middleware/check-session', function () {
...
+                steps: sessionSteps
+            });
+
+            controller.options.next = steps['/step1'].next;
+            controller.options.forks = steps['/step1'].forks;
+
+            checkProgress('/step1', controller, steps, '/');
+        });
+
+        describe('controller.getNextStep returns \'/fork1\'', function () {
```
Comment (JoeChapman) [inline test/middleware/spec.check-progress.js]: revise to 'when getNextStep returns fork1`

File: test/middleware/spec.check-progress.js [hunk 2]
```diff
@@ -50,4 +51,173 @@ describe('middleware/check-session', function () {
...
+            beforeEach(function () {
+                controller.getNextStep = sinon.stub().returns('/fork1');
+                controller.emit('complete', req, res);
+            });
+
+            it('doesn\'t remove any steps from sessionModel', function () {
```
Comment (JoeChapman) [inline test/middleware/spec.check-progress.js]: revise to 'it doesn't remove any steps' or 'no steps are removed'

File: test/middleware/spec.check-progress.js [hunk 3]
```diff
@@ -50,4 +51,173 @@ describe('middleware/check-session', function () {
         });
     });
 
+    describe('step invalidation', function () {
```
Comment (JoeChapman) [inline test/middleware/spec.check-progress.js]: Maybe revise to 'invalidates steps'

**Comment by joefitter** [comment]
@gavboulton changed the wording as per @JoeChapman 's comments. Is this what you meant?

File: test/middleware/spec.check-progress.js [hunk 4]
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
...
+                steps: sessionSteps
+            });
+
+            controller.options.next = steps['/step1'].next;
+            controller.options.forks = steps['/step1'].forks;
+
+            checkProgress('/step1', controller, steps, '/');
+        });
+
+        describe('when getNextStep returns \'/fork1\'', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: Do these tests describe a general forked journey?

File: test/middleware/spec.check-progress.js [hunk 5]
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
...
+                    'step-3-field-2',
+                    'step-4-field-1',
+                    'step-4-field-2',
+                    'fork-field-1',
+                    'fork-field-2'
+                );
+            });
+        });
+
+        describe('when getNextStep returns \'/step2\'', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: Is this when a journey no longer follows a fork?

File: test/middleware/spec.check-progress.js [hunk 6]
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
...
+          });
+
+          it('doens\'t remove step-2-field-1 or step-2-field-2', function () {
+              req.sessionModel.get('step-2-field-1').should.be.true;
+              req.sessionModel.get('step-2-field-2').should.be.true;
+          });
+
+        });
+
+        describe('when getNextStep returns \'/step4\'', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: Is this when a step can be reached as part of normal and forked journeys?

**Comment by joefitter** [comment]
Ah I think I get you @gavboulton

**Comment by joefitter** [comment]
@gavboulton @easternbloc @JoeChapman @HughePaul @daniel-ac-martin can I get a couple of 👍's?

File: test/middleware/spec.check-progress.js [hunk 7]
```diff
@@ -1,7 +1,8 @@
 var checkProgress = require('../../lib/middleware/check-progress'),
     Model = require('../../lib/model'),
     Controller = require('../../lib/controller'),
-    helpers = require('../../lib/util/helpers');
+    helpers = require('../../lib/util/helpers'),
+    _ = require('underscore');
 
 describe('middleware/check-session', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: This is an old mistake but should be 'middleware/check-progress'. I'll raise it as a separate PR.

File: test/middleware/spec.check-progress.js [hunk 8]
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
         });
     });
 
+    describe('invalidates steps', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: There are also fields and sometimes they aren't invalidated.

File: test/middleware/spec.check-progress.js [hunk 9]
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
...
+                    'step-3-field-2',
+                    'step-4-field-1',
+                    'step-4-field-2',
+                    'fork-field-1',
+                    'fork-field-2'
+                );
+            });
+        });
+
+        describe('When result of forking skips a single step', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: The tests in this block are for steps and fields relating to a fork. Have you bypassed a previous fork here?

File: test/middleware/spec.check-progress.js [hunk 10]
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
...
+
+            beforeEach(function () {
+              controller.getNextStep = sinon.stub().returns('/step2');
+              controller.emit('complete', req, res);
+            });
+
+            it('invalidates fork1', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: invalidates forked journey?

File: test/middleware/spec.check-progress.js [hunk 11]
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
...
+                req.sessionModel.get('steps').should.not.contain('/fork1');
+            });
+
+            it('invalidates fork-field-1 and fork-field-2', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: invalidates fields associated with a forked journey

File: test/middleware/spec.check-progress.js [hunk 12]
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
...
+                expect(req.sessionModel.get('fork-field-1')).to.be.undefined;
+                expect(req.sessionModel.get('fork-field-2')).to.be.undefined;
+            });
+
+            it('doens\'t invalidate step2', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: doesn't invalidate the next step in the current journey?

File: test/middleware/spec.check-progress.js [hunk 13]
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
...
+                req.sessionModel.get('steps').should.contain('/step2');
+            });
+
+            it('doens\'t remove step-2-field-1 or step-2-field-2', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: doesn't unset fields associated with the next step in the current journey?

File: test/middleware/spec.check-progress.js [hunk 14]
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
...
+        });
+
+        describe('when the result of forking skips multiple steps', function () {
+
+            beforeEach(function () {
+                controller.getNextStep = sinon.stub().returns('/step4');
+                controller.emit('complete', req, res);
+            });
+
+            it('invalidates step2, step3, and fork1', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: invalidates skipped steps?

File: test/middleware/spec.check-progress.js [hunk 15]
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
...
+                req.sessionModel.get('steps').should.not.contain.any('/step2', '/step3', '/fork1');
+            });
+
+            it('doesn\'t invalidate step1 or step4', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: doesn't invalidate steps in current journey? Or maybe new journey depending on how you look at it.

File: test/middleware/spec.check-progress.js [hunk 16]
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
...
+                req.sessionModel.get('steps').should.contain.all('/step1', '/step4');
+            });
+
+            it('invalidates all fields from /step2, /step3 and /fork1', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: invalidates all fields associated with skipped steps?

File: test/middleware/spec.check-progress.js (hunk 9 above)
Comment (joefitter) [inline test/middleware/spec.check-progress.js]: The fork is invalidated as it is unreachable with the result of getNextStep

File: test/middleware/spec.check-progress.js (hunk 10 above)
Comment (joefitter) [inline test/middleware/spec.check-progress.js]: there are 2 forked journeys - step4 and fork one. in this case only fork1 should be invalidated as step4 is still reachable

File: test/middleware/spec.check-progress.js (hunk 9 above)
Comment (joefitter) [inline test/middleware/spec.check-progress.js]: This is a 3 'pronged' fork, any outcome will technically be 'forking' - next, or either fork target

File: test/middleware/spec.check-progress.js (hunk 10 above)
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: I see.

File: test/middleware/spec.check-progress.js (hunk 9 above)
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: 'when a step is skipped'?

**Comment by joefitter** [comment]
updated @gavboulton

**Comment by joefitter** [comment]
goalposts have moved ;) - rebasing

**Comment by joefitter** [comment]
@gavboulton how is this looking now?

**Comment by gavboulton** [comment]
Looking.

File: test/middleware/spec.check-progress.js [hunk 17]
```diff
@@ -49,4 +49,164 @@ describe('middleware/check-session', function () {
...
+        });
+
+        describe('when a step is skipped', function () {
+
+            beforeEach(function () {
+                controller.getNextStep = sinon.stub().returns('/step2');
+                controller.emit('complete', req, res);
+            });
+
+            it('invalidates fork1', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: invalidates skipped step?

File: test/middleware/spec.check-progress.js [hunk 18]
```diff
@@ -49,4 +49,164 @@ describe('middleware/check-session', function () {
...
+                req.sessionModel.get('steps').should.not.contain('/fork1');
+            });
+
+            it('invalidates fields associated with fork1', function () {
```
Comment (gavboulton) [inline test/middleware/spec.check-progress.js]: invalidates fields associated with skipped step?

**Comment by joefitter** [comment]
@gavboulton thanks, updated

**Comment by gavboulton** [comment]
hmpo-form-wizard@4.4.0 published.

**Pull request merged**